pkg/session/blocklist.go

```diff
@@ -0,0 +1,163 @@
+// everest
+// Copyright (C) 2025 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package session
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/golang-jwt/jwt/v5"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	dataKey         = "list"
+	maxRetries      = 10
+	backoffInterval = 500 * time.Millisecond
+)
+
+var (
+	errExtractJti       = errors.New("could not extract jti")
+	errExtractExp       = errors.New("could not extract exp")
+	errEmptyToken       = errors.New("token is empty")
+	errUnsupportedClaim = func(claims any) error {
+		return errors.New(fmt.Sprintf("unsupported claims type: %T", claims))
+	}
+)
+
+// Blocklist represents interface to block JWT tokens and check if a token is blocked.
+type Blocklist interface {
+	// Block invalidates the token from the context by adding it to blocklist.
+	Block(ctx context.Context, token *jwt.Token) error
+	// IsBlocked checks if the token from the context is blocked.
+	IsBlocked(ctx context.Context, token *jwt.Token) (bool, error)
+}
+
+type blocklist struct {
+	tokenStore TokenStore
+	l          *zap.SugaredLogger
+}
+
+// TokenStore represents an abstraction for storage, hiding details about how the data is actually stored.
+type TokenStore interface {
+	// Add adds the shortened token to the blocklist
+	Add(ctx context.Context, shortenedToken string) error
+	// Exists checks if the shortened token is in the blocklist
+	Exists(ctx context.Context, shortenedToken string) (bool, error)
+}
+
+// BlocklistClient supports only the k8s API methods that are needed for blocklist management.
+// A separate client is needed to apply the controller-runtime cache only to the related objects.
+// Using the controller-runtime client is also beneficial because it supports HA mode.
+type BlocklistClient interface {
+	// GetSecret returns a secret that matches the criteria.
+	GetSecret(ctx context.Context, key client.ObjectKey) (*corev1.Secret, error)
+	// CreateSecret creates a secret.
+	CreateSecret(ctx context.Context, secret *corev1.Secret) (*corev1.Secret, error)
+	// UpdateSecret updates a secret.
+	UpdateSecret(ctx context.Context, secret *corev1.Secret) (*corev1.Secret, error)
+}
```

Comment:
We should move this to `token_store.go` as well since this is an implementation detail of the token store, not the blocklist. 

Also, should we name it `SecretClient` instead?

Reply:
> We should move this to token_store.go
I disagree, see https://github.com/percona/everest/pull/1293#discussion_r2063187037 

> Also, should we name it SecretClient instead?
what if for another implementation of token storage we would need another subset of methods to communicate with the k8s API? The `BlocklistClient` defines the list of whatever external methods we need to make the current implementation of the Blocklist work. Currently in includes only secrets, but we may need something else to be included and then the name `SecretClient` will get stale.

Reply:
> what if for another implementation of token storage we would need another subset of methods to communicate with the k8s API?

Then it needs to be coupled with the implementation of `TokenStore`, not `Blocklist`.. the current implementation assumes that a `TokenStore` will always be a `k8s` dependency, which defeats the purpose of introducing a `TokenStore` interface ...

Reply:
Renaming the interface to TokenStoreClient would make sense, wdyt?

Reply:
This thread https://github.com/percona/everest/pull/1293#discussion_r2063029462 changed my mind about how we handle dependencies, so moved the interface into the token store. Kept the common name TokenStoreClient without mentioning the secrets, because over the time we could not use Secrets anymore.